### PR TITLE
yr/tropicalBasics

### PR DESCRIPTION
**Pull request by YueRen**
Do not merge yet, this is just a test to see whether everything is fine.

### CODE REVIEW

File: src/TropicalGeometry/curve.jl [hunk 1]
```diff
@@ -268,22 +304,31 @@ Given a divisor `dtc` and vertex labelled `position`, compute the linearly equiv
 julia> IM = IncidenceMatrix([[1,2],[1,3],[1,4],[2,3],[2,4],[3,4]]);
 
 julia> tc = TropicalCurve(IM)
-Abstract min tropical curve
+ERROR: MethodError: no method matching TropicalCurve(::Polymake.LibPolymake.IncidenceMatrixAllocated{Polymake.LibPolymake.NonSymmetric})
```
Comment (fingolfin) [inline src/TropicalGeometry/curve.jl]: Error here and below "intentional" (or rather: known and expected)?

Comment (YueRen) [reply src/TropicalGeometry/curve.jl]: No, the erros are not known but also not expected.  Let me just quickly check with some people downstairs in Paderborn.

Comment (fingolfin) [reply src/TropicalGeometry/curve.jl]: Any progress here, or plans for progress, or things others could help with?

Comment (YueRen) [reply src/TropicalGeometry/curve.jl]: Not yet. I am currently polishing the OSCAR chapter and this together, based on the reviews that I have received.

**Comment by YueRen** [comment]
Some of the doctest is failing because OSCAR is printing out a set and the other of the element is different.  Is there a way to take these documentations out of the doctest?

**Comment by fieker** [comment]
On Mon, Oct 09, 2023 at 03:19:23AM -0700, YueRen wrote:
> Some of the doctest is failing because OSCAR is printing out a set and the other of the element is different.  Is there a way to take these documentations out of the doctest?
> 
Yep. You can add regular expressions to add and ignore

DocTestFilters = r"Group\(.*\]\)"

e.g. in docs/src/NumberTheory/galois.md

this removed the comparison of groups as the generators are randomized

File: src/TropicalGeometry/curve.jl [hunk 2]
```diff
@@ -1,536 +1,549 @@
...
-"""
-coefficients(dtc::DivisorOnTropicalCurve) = dtc.coefficients
-
-@doc raw"""
-   degree(dtc::DivisorOnTropicalCurve)
-
-Compute the degree of  a divisor `dtc` on an abstract tropical curve.
-
-# Examples
 ```jldoctest
```
Comment (fingolfin) [inline src/TropicalGeometry/curve.jl]: Instead do
````suggestion
```jldoctest; filter = r"Edge\(.*\)"
````

see also https://documenter.juliadocs.org/stable/man/doctests/#Filtering-Doctests

File: src/TropicalGeometry/curve.jl [hunk 3]
```diff
@@ -1,536 +1,549 @@
...
-DivisorOnTropicalCurve{min, false}(Abstract min tropical curve, [0, 1, 1, 1])
-
-julia> coefficients(dtc)
-4-element Vector{Int64}:
- 0
- 1
- 1
- 1
+```@meta
+DocTestFilters = r"Edge\(.*\)"
```
Comment (fingolfin) [inline src/TropicalGeometry/curve.jl]: Don't do this for a single doctest...

**Comment by YueRen** [comment]
What do I have to do in order to get the tests running on github.  I want to check whether there are any more failing tests.

**Comment by HereAround** [comment]
The tests will run once you fix the merge conflicts.

**Comment by YueRen** [comment]
So I will need to rebase it again?  But how can there be merge conflicts? Did somebody change those files in the tropical folder in the last 48h?

**Comment by HereAround** [comment]
That is one way to fix it (and how I would do it). Presumably, you can also hit "resolve conflicts" below, but I have never done that before.

**Comment by HereAround** [comment]
> So I will need to rebase it again? But how can there be merge conflicts? Did somebody change those files in the tropical folder in the last 48h?

The conflicts are in the following files:

```julia
 src/TropicalGeometry/initial.jl
src/TropicalGeometry/linear_space.jl
src/TropicalGeometry/valuation.jl
src/TropicalGeometry/variety.jl
```
The last commit that edited the file `variety.jl` is the following:
```
commit d15d9a8e0ac38d9707749c4aa3fb46812637e834
Author: Max Horn <max@quendi.de>
Date:   Tue Oct 10 08:21:16 2023 +0200

    Bump dependencies, adjust to changes (#2903)
    
    * Bump dependencies, adjust to changes
    
    * rename `mat` to `matrix`
    
    ---------
    
    Co-authored-by: Felix Röhrich <roehrich@art.rwth-aachen.de>
```
(In case you wonder, I found this via `git log -p <filepath>`.)

**Comment by YueRen** [comment]
Got it, I was just majorly confused.

**Comment by YueRen** [comment]
There are still some tests failing, but I don't know why from the error messages.  Can somebody give me a hint if there's still something wrong with my code?  Otherwise, apart from some cosmetic changes the pull-request is ready to go.

**Comment by benlorenz** [comment]
> There are still some tests failing, but I don't know why from the error messages. Can somebody give me a hint if there's still something wrong with my code? Otherwise, apart from some cosmetic changes the pull-request is ready to go.

It is still marked as a draft.
A large chunk of the file `src/TropicalGeometry/curve.jl` is commented, in particular the statement `using RecipesBase` which causes Aqua.jl to fail because we now have a unused dependency. What is the plan with this code? (There are large commented code blocks in other files as well)

**Comment by lgoettgens** [comment]
I think you removed the last instance of an `using RecipesBase` from Oscar in `src/TropicalGeometry/curve.jl:6`. So we no longer need that package as a dependency, and it can be thus removed from `Project.toml`.

**Comment by YueRen** [comment]
I keep getting the following error in the nightly test, is this caused by my changes? I haven't made any changes to `runtests.jl`.

```
[ Info: /home/runner/work/Oscar.jl/Oscar.jl/test/runtests.jl -- fixed SEED 42
ERROR: LoadError: TaskFailedException

    nested task error: 
GC: pause 148.68ms. collected 388.283483MB. incr 
Heap stats: bytes_mapped 512.12 MB, bytes_resident 323.12 MB,
heap_size 510.01 MB, heap_target 597.51 MB, Fragmentation 0.683
invalid method definition in Main: function Main.include must be explicitly imported to be extended
    Stacktrace:
     [1] top-level scope
       @ none:0
     [2] top-level scope
       @ none:1
     [3] eval
       @ Core ./boot.jl:398 [inlined]
     [4] (::Distributed.var"#172#174"{Module, Expr})()
       @ Distributed /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Distributed/src/macros.jl:233
Stacktrace:
 [1] sync_end(c::Channel{Any})
   @ Base ./task.jl:450
 [2] macro expansion
   @ ./task.jl:483 [inlined]
 [3] remotecall_eval(m::Module, procs::Vector{Int64}, ex::Expr)
   @ Distributed /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Distributed/src/macros.jl:219
 [4] top-level scope
   @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Distributed/src/macros.jl:203
 [5] include(fname::String)
   @ Base.MainInclude ./client.jl:495
 [6] top-level scope
   @ none:6
in expression starting at /home/runner/work/Oscar.jl/Oscar.jl/test/runtests.jl:85
ERROR: LoadError: Package Oscar errored during testing
Stacktrace:
 [1] pkgerror(msg::String)
   @ Pkg.Types /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Pkg/src/Types.jl:70
 [2] test(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; coverage::Bool, julia_args::Cmd, test_args::Cmd, test_fn::Nothing, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool)
   @ Pkg.Operations /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Pkg/src/Operations.jl:2020
 [3] test
   @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Pkg/src/Operations.jl:1901 [inlined]
 [4] test(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; coverage::Bool, test_fn::Nothing, julia_args::Vector{String}, test_args::Cmd, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool, kwargs::@Kwargs{io::Pkg.UnstableIO})
   @ Pkg.API /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Pkg/src/API.jl:444
 [5] test(pkgs::Vector{Pkg.Types.PackageSpec}; io::Pkg.UnstableIO, kwargs::@Kwargs{coverage::Bool, julia_args::Vector{String}, force_latest_compatible_version::Bool})
   @ Pkg.API /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Pkg/src/API.jl:159
 [6] test(; name::Nothing, uuid::Nothing, version::Nothing, url::Nothing, rev::Nothing, path::Nothing, mode::Pkg.Types.PackageMode, subdir::Nothing, kwargs::@Kwargs{coverage::Bool, julia_args::Vector{String}, force_latest_compatible_version::Bool})
   @ Pkg.API /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.11/Pkg/src/API.jl:174
 [7] top-level scope
   @ ~/work/_actions/julia-actions/julia-runtest/latest/test_harness.jl:15
 [8] include(fname::String)
   @ Base.MainInclude ./client.jl:495
 [9] top-level scope
   @ none:1
in expression starting at /home/runner/work/_actions/julia-actions/julia-runtest/latest/test_harness.jl:7
Error: Process completed with exit code 1.
```

**Comment by thofma** [comment]
I don't think it is your fault. @benlorenz any idea?

**Comment by benlorenz** [comment]
> I keep getting the following error in the nightly test, is this caused by my changes? I haven't made any changes to `runtests.jl`.
> 
> ```
> invalid method definition in Main: function Main.include must be explicitly imported to be extended
> ```

This error is triggered by some recent change in julia nightly (and a hack currently in our test-driver), but nightly would fail anyway later during the tests.
The hack will be removed with #2810 which should fix that error.

**Comment by thofma** [comment]
Thanks for the explanation. @YueRen ignore 1.10 (`Error: The operation was canceled`) and nightly.

**Comment by YueRen** [comment]
@bschroter @lkastner The pull request is ready to be reviewed.  If you want, I can give you a brief rundown over Zoom to help you get started (maybe around the OSCAR meeting this Friday?).

File: Project.toml [hunk 4]
```diff
@@ -40,8 +39,6 @@ Polymake = "0.11.6"
 Preferences = "1"
 Random = "1.6"
 RandomExtensions = "0.4.3"
-RecipesBase = "1.2.1"
-Serialization = "1.6"
 Singular = "0.18.17"
```
Comment (benlorenz) [inline Project.toml]:
```suggestion
Serialization = "1.6"
Singular = "0.18.17"
```
Please don't remove the compat entry for Serialization.

File: src/exports.jl [hunk 5]
```diff
@@ -1489,3 +1478,13 @@ export write_as_full
...
+
+# for tropical_bacics:
+export algebraic_ideal
+export algebraic_matrix
+export algebraic_polynomial
+export algebraic_pluecker_vector
+export pluecker_indices
+export tropical_pluecker_vector
+export tropical_matrix
+# export tropical_polynomial # already exported
```
Comment (benlorenz) [inline src/exports.jl]: Please insert them properly.

File: docs/src/TropicalGeometry/groebner_theory.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# Groebner theory
+
+## Introduction
+Groebner bases and related notions in tropical geometry are a generalization of its classical counterparts to fields with valuation.  The Groebner bases take the valuation into account, the initial ideals live over the residue field, and consequently the Gr\"obner complex is not necessarily a polyhedral fan.  For details, see
```
Comment (benlorenz) [inline docs/src/TropicalGeometry/groebner_theory.md]:
```suggestion
Groebner bases and related notions in tropical geometry are a generalization of its classical counterparts to fields with valuation.  The Groebner bases take the valuation into account, the initial ideals live over the residue field, and consequently the Groebner complex is not necessarily a polyhedral fan.  For details, see
```

File: docs/src/TropicalGeometry/intro.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+- [MS15](@cite)
+- [Jos21](@cite)
+
+Like [MS15](@cite), OSCAR follows the min-convention by default and unless specified otherwise.
+
+## Contact
+Please direct questions about this part of OSCAR to:
+[//]: <> * [Lars Kastner](https://lkastner.github.io/),
+[comment]: <> * [Marta Panizzut](https://martapanizzut.github.io/),
+* [Yue Ren](https://www.yueren.de/).
```
Comment (benlorenz) [inline docs/src/TropicalGeometry/intro.md]: This [looks](https://docs.oscar-system.org/previews/PR2832/TropicalGeometry/intro/#Contact) weird.

File: docs/src/TropicalGeometry/misc.md
Comment (benlorenz) [inline docs/src/TropicalGeometry/misc.md:1]: I guess this file can be removed? It is not included and basically empty.

File: docs/src/TropicalGeometry/semiring.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+Whereas tropical semirings in [MS15](@cite) and [Jos21](@cite) are extensions of the real numbers, tropical semirings in OSCAR are an extension of the rational numbers to avoid precision issues.
+
+## Constructor
+Objects of type `TropicalSemiring`, as well as matrices and polynomials thereover, can be constructed as follows:
+```@docs
+tropical_semiring()
+```
+
+## Properties
+Objects of type `TroicalSemiring` have the following properties:
```
Comment (benlorenz) [inline docs/src/TropicalGeometry/semiring.md]:
```suggestion
Objects of type `TropicalSemiring` have the following properties:
```

File: src/Combinatorics/Matroids/properties.jl
Comment (benlorenz) [inline src/Combinatorics/Matroids/properties.jl:1]: Please undo the changes in this file.

File: src/TropicalGeometry/semiring.jl [hunk 9]
```diff
@@ -81,157 +113,158 @@ false
...
-julia> T = TropicalSemiring()
-Tropical semiring (min)
+julia> T = tropical_semiring()
+Min tropical semiring
 
-julia> A = [T(0) inf(T); inf(T) T(0)] # = tropical identity matrix
-2×2 Matrix{Oscar.TropicalSemiringElem{typeof(min)}}:
- (0)  ∞
- ∞    (0)
+julia> A = matrix(T,[0 zero(T); zero(T) 0]) # = tropical identity matrix
```
Comment (benlorenz) [inline src/TropicalGeometry/semiring.jl]:
```suggestion
julia> A = identity_matrix(T, 2) # = tropical identity matrix
```

File: src/TropicalGeometry/semiring.jl [hunk 10]
```diff
@@ -81,157 +113,158 @@ false
...
-#  Predicates
+#  Printing
 #
 ################################################################################
 
-Oscar.iszero(x::TropicalSemiringElem) = isinf(x)
+# Hook into the fancy printing, we use (x) for finite values and ±∞ for infinity.
+function AbstractAlgebra.expressify(x::TropicalSemiringElem{minOrMax}; context = nothing) where {minOrMax<:Union{typeof(min),typeof(max)}}
+  if isinf(x)
+    return minOrMax==typeof(min) ? "∞" : "-∞"
```
Comment (benlorenz) [inline src/TropicalGeometry/semiring.jl]: This should probably use `Oscar.is_unicode_allowed()` and print `inf` / `-inf` otherwise.

File: src/TropicalGeometry/semiring.jl [hunk 11]
```diff
@@ -81,157 +113,158 @@ false
...
-# Test if something is inf.
-isinf(x::TropicalSemiringElem) = x.isinf
+function (R::TropicalSemiring)(x::Union{Integer, Rational}; preserve_ordering::Bool=false)
+  return R(QQ(x),preserve_ordering=preserve_ordering)
+end
 
-Oscar.parent(x::TropicalSemiringElem{T}) where T = TropicalSemiring{T}()
+function (R::TropicalSemiring)(x::RingElem; preserve_ordering::Bool=false)
+  x = QQ(x)
+  @assert parent(x)==QQ "cannot convert object of type $(repr(typeof(x)))"
```
Comment (benlorenz) [inline src/TropicalGeometry/semiring.jl]:
```suggestion
  @req parent(x)==QQ "cannot convert object of type $(repr(typeof(x)))"
```
Please use `@req` instead of `@assert`.

File: src/TropicalGeometry/semiring.jl [hunk 12]
```diff
@@ -81,157 +113,158 @@ false
...
+  return R(x, preserve_ordering=preserve_ordering)
+end
 
-@doc raw"""
-    convention(T::TropicalSemiring)
+function (::QQField)(x::TropicalSemiringElem{typeof(min)}; preserve_ordering::Bool=false)
+    @assert !iszero(x) "cannot convert $(repr(x))"
```
Comment (benlorenz) [inline src/TropicalGeometry/semiring.jl]:
```suggestion
    @req !iszero(x) "cannot convert $(repr(x))"
```

File: src/TropicalGeometry/semiring.jl [hunk 13]
```diff
@@ -81,157 +113,158 @@ false
...
+    return data(x)
+end
 
-Return `min` if `T` is the min tropical semiring,
-return `max` if `T` is the max tropical semiring.
+function (::QQField)(x::TropicalSemiringElem{typeof(max)}; preserve_ordering::Bool=false)
+    @assert !iszero(x) "cannot convert $(repr(x))"
```
Comment (benlorenz) [inline src/TropicalGeometry/semiring.jl]:
```suggestion
    @req !iszero(x) "cannot convert $(repr(x))"
```

File: src/TropicalGeometry/semiring.jl [hunk 14]
```diff
@@ -247,376 +280,115 @@ end
...
-  for s in S
-    o = one(R)
-    for i in 1:nrows(x)
-      o = o * x[i, s[i]]
-    end
-    res = res + o
-  end
-  return res
+function divexact(x::TropicalSemiringElem{minOrMax}, y::TropicalSemiringElem{minOrMax}) where {minOrMax<:Union{typeof(min),typeof(max)}}
+    @assert !iszero(y) "dividing by (tropical) zero"
```
Comment (benlorenz) [inline src/TropicalGeometry/semiring.jl]:
```suggestion
    @req !iszero(y) "dividing by (tropical) zero"
```

File: src/TropicalGeometry/semiring.jl [hunk 15]
```diff
@@ -247,376 +280,115 @@ end
...
-         push!(mins, determinant(A[ri, ci]))
-      end
-   end
-   return(mins)
+function Base.:(/)(x::TropicalSemiringElem{minOrMax}, y::TropicalSemiringElem{minOrMax}) where {minOrMax<:Union{typeof(min),typeof(max)}}
+    return divexact(x,y)
 end
 
+function inv(a::TropicalSemiringElem)
+    @assert !iszero(a) "inverting (tropical) zero"
```
Comment (benlorenz) [inline src/TropicalGeometry/semiring.jl]:
```suggestion
    @req !iszero(a) "inverting (tropical) zero"
```

File: src/TropicalGeometry/semiring.jl [hunk 16]
```diff
@@ -247,376 +280,115 @@ end
...
+    return parent(a)(-data(a))
+end
 
+function Base.:(^)(a::TropicalSemiringElem, n::QQFieldElem)
```
Comment (benlorenz) [inline src/TropicalGeometry/semiring.jl:369]: This and the following should probably error `iszero(a)` and the exponent is negative?

```
julia> t = zero(T);

julia> t^-1
∞
```

File: src/TropicalGeometry/semiring_map.jl [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+
+julia> nu_t(1//t^2)
+(2)
+
+```
+"""
+function tropical_semiring_map(Kt::Generic.RationalFunctionField, t::Generic.RationalFunctionFieldElem, minOrMax::Union{typeof(min),typeof(max)}=min)
+    @assert isone(denominator(t)) "input uniformizer denominator not 1"
+    t = numerator(t)
+    @assert degree(t)>0 "input uniformizer constant"
```
Comment (benlorenz) [inline src/TropicalGeometry/semiring_map.jl]:
```suggestion
    @req isone(denominator(t)) "input uniformizer denominator not 1"
    t = numerator(t)
    @req degree(t)>0 "input uniformizer constant"
```

File: src/TropicalGeometry/TropicalGeometry.jl [hunk 18]
```diff
@@ -1,54 +1,34 @@
...
-
-
-###
-# Allow gcd of vectors of univariate rational polynomials
-# to make their handling similar to that of integers
-###
-gcd(F::Vector{QQPolyRingElem}) = reduce(gcd, F)
-gcd(F::QQPolyRingElem...) = reduce(gcd, F)
+# Workaround of issue https://github.com/oscar-system/Oscar.jl/issues/2222
+Base.hash(x::Polyhedron, h::UInt) = h
```
Comment (benlorenz) [inline src/TropicalGeometry/TropicalGeometry.jl]: I have added this in #2955, so this can be removed once the other PR is merged.
```suggestion
```

File: src/TropicalGeometry/curve.jl [hunk 19]
```diff
@@ -1,536 +1,549 @@
...
+#  Tropical curves
+#  ===============
+#  concrete subtype of TropicalVarietySupertype in variety_supertype.jl
+#
+###############################################################################
+
+@attributes mutable struct TropicalCurve{minOrMax,isEmbedded} <: TropicalVarietySupertype{minOrMax,isEmbedded}
+    polyhedralComplex::Union{PolyhedralComplex,
+                             Graph}
+    multiplicities::Union{Dict{<:Polyhedron,ZZRingElem},
```
Comment (benlorenz) [inline src/TropicalGeometry/curve.jl]: I would strongly recommend not using the polyhedra as key in a Dict, while this  does work with the hash now it is still extremely expensive to compare polytopes (it needs a primal and a dual description for each polytope).

File: src/TropicalGeometry/variety.jl [hunk 20]
```diff
@@ -1,390 +1,840 @@
...
-function Base.show(io::IO, tv::TropicalVariety{M, EMB}) where {M, EMB}
-    if EMB
-        print(io, "$(repr(M)) tropical variety of dimension $(dim(tv)) embedded in $(ambient_dim(tv))-dimensional Euclidean space")
-    else
-        print(io, "Abstract $(repr(M)) tropical variety of dimension $(dim(tv))")
+    return TropV
+end
+
+
+function polyhedral_complex(Sigma::Vector{<:Polyhedron})
```
Comment (benlorenz) [inline src/TropicalGeometry/variety.jl]: I will provide a PR which does this properly.

**Review by benlorenz — COMMENTED** [review]
A first round of comments.

File: src/TropicalGeometry/curve.jl (hunk 19 above)
Comment (YueRen) [reply src/TropicalGeometry/curve.jl]: What better options are there?  

The canonical option (for me) would be to store all multiplicities in a vector, but there are two questions:
1. Is the order of `maximal_polyhedra(::PolyhedralComplex)` fixed once the polyhedral complex has been constructed (i.e., will not change regardless what I do with the polyhedral complex)?
2. Does the order of `maximal_polyhedra` coincide with the order of the `IncidenceMatrix` with which I construct the `PolyhedralComplex`?  If no, is there a cheap way to find out which entry of `maximal_polyhedra` correspond to which row of the `IncidenceMatrix`?

File: src/TropicalGeometry/variety.jl (hunk 20 above)
Comment (YueRen) [reply src/TropicalGeometry/variety.jl]: That's great!  If you do, can you please add an option to disable the checks whether the input polyhedra actually form a polyhedral complex?  I should always be calling it with a set of maximal polyhedra of a polyhedral complex, no checks required.

File: docs/src/TropicalGeometry/intro.md (hunk 7 above)
Comment (YueRen) [reply docs/src/TropicalGeometry/intro.md]: Fixed, thanks for catching that!

File: src/exports.jl (hunk 5 above)
Comment (YueRen) [reply src/exports.jl]: Fixed, thanks for pointing that out!

File: docs/src/TropicalGeometry/misc.md
Comment (YueRen) [reply docs/src/TropicalGeometry/misc.md:1]: Done, thanks for catching that!

File: src/TropicalGeometry/curve.jl (hunk 19 above)
Comment (benlorenz) [reply src/TropicalGeometry/curve.jl]: 1. Once you access the `maximal_polyhedra` once it will stay fixed (corresponding to the rows of `MAXIMAL_POLYTOPES` in the polymake object)
2. If you construct the complex from an `IncidenceMatrix` with the option `non_redundant=true` then this matrix will be used for `MAXIMAL_POLYTOPES` and `maximal_polyhedra` will have the same order.
   (But make sure this is properly canonicalized, e.g. each vertex only appears once.)

You can check for permutations between incidence matrices like this:
```
julia> m1 = maximal_polyhedra(IncidenceMatrix, pc)
2×7 IncidenceMatrix
[1, 2, 3, 4]
[4, 5, 6, 7]


julia> Polymake.common.find_permutation(Polymake.common.rows(m1), Polymake.common.rows(m1))
pm::Array<long>
0 1
```
this returns a 0-based permutation between the rows of two incidence matrices.

File: src/TropicalGeometry/semiring.jl (hunk 16 above)
Comment (YueRen) [reply src/TropicalGeometry/semiring.jl:369]: Correct, it should be fixed now.

File: src/TropicalGeometry/semiring.jl (hunk 10 above)
Comment (YueRen) [reply src/TropicalGeometry/semiring.jl]: Fixed, thanks for pointing that out.  Didn't know that unicode was disabled by default.

**Comment by YueRen** [comment]
Some of the tests that are currently failing are in the matroid package.  I don't think they are related to my changes.

File: src/Combinatorics/Matroids/properties.jl [hunk 21]
```diff
@@ -161,7 +161,7 @@ function flats_impl(M::Matroid, r::Union{Int,Nothing}, num_flats::Int, pm_flats)
         jl_flats = reverse(jl_flats)
     end
     matroid_flats = [[M.groundset[i] for i in flat] for flat in jl_flats]
-    if r !== nothing
+    if r!=nothing
```
Comment (benlorenz) [inline src/Combinatorics/Matroids/properties.jl]:
```suggestion
    if r !== nothing
```
Please undo the changes in this file, this should fix the CI errors. (And maybe merge master into your branch)

File: src/Combinatorics/Matroids/properties.jl [hunk 22]
```diff
@@ -854,7 +854,7 @@ julia> tutte_connectivity(fano_matroid())
 3
 
 julia> tutte_connectivity(uniform_matroid(2,4))
-infinity
+PosInf()
```
Comment (benlorenz) [inline src/Combinatorics/Matroids/properties.jl]:
```suggestion
infinity
```

**Comment by YueRen** [comment]
All required tests are now running through, let me know if you need any help with the reviews @bschroter @lkastner.  I'd really like to get this merged sooner than later so my other collaborators can start building on top of it.

**Comment by fingolfin** [comment]
Unfortunately this has a merge conflict now :-/

**Comment by YueRen** [comment]
> Unfortunately this has a merge conflict now :-/

It doesn't look too hard to fix.  I will get to it once the reviews are done.

**Review by bschroter — COMMENTED** [review]
I had a quick look at the tropical functionality (mainly from a users point of view and I only briefly looked at the code). 

1) There are some larger parts of code just as a comment (see line 570 of variety.jl)
Maybe this requires first a little bit of clean up.

2) Also it says `tropical_variety is currently under development' does that mean this should be an experimental package for the moment?

3) It is unclear to me whether the debug files should be part of the code base or not.

4) Should be functionality as Gröbner fans really be hidden in the tropical source code?

5) It might be confusing that tropical_variety returns an array while the other constructors as tropical_linear_space or tropical_hypersurface, tropical_curve return the variety directly.

6) [maybe more an enhancement] Should it be possible to intersect a tropical_linear_space with a tropical_hypersurface etc.?

**Comment by YueRen** [comment]
Thanks for the comments @bschroter. 

>     1. There are some larger parts of code just as a comment (see line 570 of variety.jl)
>        Maybe this requires first a little bit of clean up.

The code is kind of important, it is the beginning of the code for computing arbitrary tropicalizations.  However, there are still a couple of hard-to-fix bugs that need to be ironed out.

>     2. Also it says `tropical_variety is currently under development' does that mean this should be an experimental package for the moment?

The functionality that is there works or should work.  But it is not feature complete yet.

>     4. Should be functionality as Gröbner fans really be hidden in the tropical source code?

No, it should be documented.  I'll see whether I can add some quick documentation for it.

>     5. It might be confusing that tropical_variety returns an array while the other constructors as tropical_linear_space or tropical_hypersurface, tropical_curve return the variety directly.

True, maybe `tropical_variety` should just assume that the input is a prime ideal then and introduce a new function `factorize_and_tropicalize` or something similar.  Or alternatively rename the function to `tropical_varieties` instead.

>     6. [maybe more an enhancement] Should it be possible to intersect a tropical_linear_space with a tropical_hypersurface etc.?

If there is a quick algorithm for doing so, sure!  Otherwise, their intersection should fall back to the general intersection of tropical varieties.

**Comment by YueRen** [comment]
@lkastner @benlorenz @thofma Some tests in PolyhedralGeometry are failing for me, do you know what may be causing them?
https://github.com/oscar-system/Oscar.jl/actions/runs/6982130645/job/19000676739?pr=2832

It looks Nemo related to me.

**Comment by thofma** [comment]
Without being at a computer, I am not aware of any change that might have caused this. @benlorenz did something change on the polyhedral side?

**Comment by benlorenz** [comment]
The backtrace for the failing polyhedral tests contains:
```
 Oscar ~/work/Oscar.jl/Oscar.jl/src/TropicalGeometry/variety.jl:122
```

You need to remove the custom `polyhedral_complex(Vector{Polyhedron})` function (https://github.com/oscar-system/Oscar.jl/pull/2832/files#diff-0379b0f2b166f68b219d831526ab1e7c3e0894abb0f5db4c14b1cadb509e0ed1R109), this overrides the functions that I added recently but fails for non-rational objects.

See https://github.com/oscar-system/Oscar.jl/pull/2959#issuecomment-1808104323 for details on the functions I added.

**Comment by YueRen** [comment]
Okay, all issues seem to be fixed.  The pull request is ready to be merged as far as I am concerned.
